## Exercise 1

### Exercise 1.1

In [1]:
deckcards=[j for i in range(1,5) for j in range(1,14)]      #generating the deck of cards using list comprehension 
print(deckcards)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]


### Exercise 1.2

In [2]:
total=sum(deckcards)                                                        #sum of the list deckcards
print("Total value associated with a full deck of cards:",total)


Total value associated with a full deck of cards: 364


## Exercise 2

### Exercise 2.1

In [3]:
import numpy as np                                           #import numpy
cards=[]
player1=[]                                                  # initialize list of cards for player1
player2=[]                                                  # initialize list of cards for player1
i=0
def deal():                                                  # defining the function deal
    global cards,i
    global player1
    global player2
    cards=[j for i in range(1,5) for j in range(1,14)]       # generating the deck of cards using list comprehension 
    np.random.shuffle(cards)                                 # using np.random.shuffle to shuffle the cards 
    while i <len(cards):                                        # using while loop to iterate through cards list
        player1.append(cards[i])                                # handing out cards to players alternately 
        if (i+1<len(cards)):
            player2.append(cards[i+1])  
        i=i+2
    hand=(player1,player2)
    print("Deck of cards after shuffle",cards)
    print("player1: ",player1)
    print("player2: ",player2)
    print("Hand of players:",hand)
    return hand
deal()                                                       # calling the function deal

Deck of cards after shuffle [12, 13, 2, 8, 8, 9, 11, 4, 3, 6, 9, 13, 11, 7, 3, 12, 13, 11, 5, 7, 2, 8, 3, 4, 7, 6, 2, 8, 6, 1, 4, 13, 4, 5, 11, 3, 7, 10, 1, 10, 5, 2, 12, 9, 5, 1, 1, 10, 12, 10, 9, 6]
player1:  [12, 2, 8, 11, 3, 9, 11, 3, 13, 5, 2, 3, 7, 2, 6, 4, 4, 11, 7, 1, 5, 12, 5, 1, 12, 9]
player2:  [13, 8, 9, 4, 6, 13, 7, 12, 11, 7, 8, 4, 6, 8, 1, 13, 5, 3, 10, 10, 2, 9, 1, 10, 10, 6]
Hand of players: ([12, 2, 8, 11, 3, 9, 11, 3, 13, 5, 2, 3, 7, 2, 6, 4, 4, 11, 7, 1, 5, 12, 5, 1, 12, 9], [13, 8, 9, 4, 6, 13, 7, 12, 11, 7, 8, 4, 6, 8, 1, 13, 5, 3, 10, 10, 2, 9, 1, 10, 10, 6])


([12,
  2,
  8,
  11,
  3,
  9,
  11,
  3,
  13,
  5,
  2,
  3,
  7,
  2,
  6,
  4,
  4,
  11,
  7,
  1,
  5,
  12,
  5,
  1,
  12,
  9],
 [13,
  8,
  9,
  4,
  6,
  13,
  7,
  12,
  11,
  7,
  8,
  4,
  6,
  8,
  1,
  13,
  5,
  3,
  10,
  10,
  2,
  9,
  1,
  10,
  10,
  6])

### Exercise 2.2

In [4]:

#alternate method providing first 26 cards to player1 and other 26 to player2
player1_alt=cards[:26]
player2_alt=cards[26:]


print("Values using distributing alternate cards to player1 and player2")
print("Total number of cards with Player1: ",len(player1))
print("Total number of cards with Player2: ",len(player2))
print("Cards with player1: ",player1)
print("Cards with player2: ",player2)
print("Total value of cards of Player1: ",sum(player1))
print("Total value of cards of Player2: ",sum(player2))
print()
print("Values using alternate method of distributing first 26 cards to player1 and other 26 to player2")
print("Total number of cards with Player1: ",len(player1_alt))
print("Total number of cards with Player2: ",len(player2_alt))
print("Cards with player1: ",player1_alt)
print("Cards with player2: ",player2_alt)
print("Total value of cards of Player1: ",sum(player1_alt))
print("Total value of cards of Player2: ",sum(player2_alt))



Values using distributing alternate cards to player1 and player2
Total number of cards with Player1:  26
Total number of cards with Player2:  26
Cards with player1:  [12, 2, 8, 11, 3, 9, 11, 3, 13, 5, 2, 3, 7, 2, 6, 4, 4, 11, 7, 1, 5, 12, 5, 1, 12, 9]
Cards with player2:  [13, 8, 9, 4, 6, 13, 7, 12, 11, 7, 8, 4, 6, 8, 1, 13, 5, 3, 10, 10, 2, 9, 1, 10, 10, 6]
Total value of cards of Player1:  168
Total value of cards of Player2:  196

Values using alternate method of distributing first 26 cards to player1 and other 26 to player2
Total number of cards with Player1:  26
Total number of cards with Player2:  26
Cards with player1:  [12, 13, 2, 8, 8, 9, 11, 4, 3, 6, 9, 13, 11, 7, 3, 12, 13, 11, 5, 7, 2, 8, 3, 4, 7, 6]
Cards with player2:  [2, 8, 6, 1, 4, 13, 4, 5, 11, 3, 7, 10, 1, 10, 5, 2, 12, 9, 5, 1, 1, 10, 12, 10, 9, 6]
Total value of cards of Player1:  197
Total value of cards of Player2:  167


## Exercise 3

### Exercise 3.1

1. Shuffle deck
2. Deal half of deck to each player. Each player has a discardpile.
3. Each player plays two card at a time.
4. The player with higher sum of the two cards put all four cards (two of opponent and two of player) in his/her discardpile.
4. If there is a tie (sum of two cards is equal for both players) then, both players play next two cards.
5. Now, when sum of cards played by one player (along with the tie cards) is greater than the other players sum, the player        with greater sum put all the cards (his and opponents) in his/her discardpile. 
6. If the deck is exhausted in the middle of a tie then put all the tied cards of the respective players in their respective      discardpiles.
7. Game terminates after a maximum of 13 rounds (occurs when there are no ties) and the deck is exhausted.
8. The player with the highest total card points (in their discard pile) wins the game.

### Exercise 3.2

In [5]:
discardpile1=[]                                          # initialize list of discardpile1 for player1
discardpile2=[]                                          # initialize list of discardpile2 for player2

def play():                                              # define the function play
    i=0    

    while i<len(player1):                                # using while loop to iterate through players cards
        sum1=0                                           # initialize sum1 to 0, sum1 is sum of 2 cards played by player1
        sum2=0                                           # initialize sum2 to 0, sum1 is sum of 2 cards played by player2
        p1=[]                                            # initialize p1 i.e cards played by player1 
        p2=[]                                            # initialize p2 i.e cards played by player2
        p1.append(player1[i])                            # append cards to both lists
        p1.append(player1[i+1])
        p2.append(player2[i])
        p2.append(player2[i+1])
        sum1=sum(p1)                                        #sum to find which player has greater sum of cards
        sum2=sum(p2)
        if(sum1>sum2):                                      # check if sum1 is greater than sum2
            discardpile1.extend(p1)                         # add cards to discardpile of player1
            discardpile1.extend(p2)
            p1=[]                                           #empty p1 and p2
            p2=[]
            i=i+2                                           #increment i by 2
        elif(sum2>sum1):                                    # check if sum2 is greater than sum1
            discardpile2.extend(p1)                         # add cards to discardpile of player2
            discardpile2.extend(p2)
            p1=[]                                           # empty p1 and p2
            p2=[] 
            i=i+2                                           # increment i by 2 
        elif(sum1==sum2):                                   # else if sum1 is equal to sum2
            while(sum1==sum2 and i+2<len(player1)):         # while loop with condition the player have cards and sum is equal for both players       
                i=i+2                                       # increment i by 2        
                p1.append(player1[i])                       # append and store respective players cards in two list until tie breaks or players cards exhaust 
                p1.append(player1[i+1])
                p2.append(player2[i])
                p2.append(player2[i+1])
                sum1=sum1+sum(p1)                           # add sum of new cards to existing sum1
                sum2=sum2+sum(p2)                           # add sum of new cards to existing sum2
            i=i+2                                            # increment i by 2                          
            if(sum1>sum2):                                   # check if player1 has greater final score with respect to player2, player1 won the round    
                discardpile1.extend(p1)                      # empty p1 list to discardpile of player1
                discardpile1.extend(p2)                      # empty p2 list to discardpile of player1
                p1=[]                                         # empty p1 and p2
                p2=[]
            else:
                discardpile2.extend(p1)
                discardpile2.extend(p2)
                p1=[]
                p2=[]
    #print("final discardpile 1",discardpile1)
    #print("final discardpile 1",discardpile2)
    score1=sum(discardpile1)
    score2=sum(discardpile2)
    winner_Score=0
    winner=""
    loser_Score=0
    if(score1>score2):                                    # if final score of player1 is greater than that of player2, player1 is winner
        winner="Winner is Player1"
        print(winner)         
        print("Winner score: ",score1)
        winner_Score=score1
        print("Loser score: ",score2)
        loser_Score=score2
    elif(score2>score1):                                  # if final score of player2 is greater than that of player1, player2 is winner
        winner="Winner is Player 2"
        print(winner)
        print("Winner score: ",score2)
        winner_Score=score2
        print("Loser score: ",score1)
        loser_Score=score1
    else:                                                 # if final score of player 1 is equal to that of player2, there is a tie
        winner="There is a tie"
        print(winner)
        print("Tie score: ",score2)  
    
    return winner, winner_Score, loser_Score




### Exercise 3.3

In [6]:
play()

Winner is Player 2
Winner score:  244
Loser score:  120


('Winner is Player 2', 244, 120)

### Exercise 4.1

In [7]:
import statistics 
initial_total1=[]                                                 # list of initial totals of player1
initial_total2=[]                                                 # list of initial totals of player2
win_player1=[0 for i in range(20)]                                # list to determine which matches were won by player1
win_player2=[0 for i in range(20)]                                # list to determine which matches were won by player2
win=[]                                                            # final scores of winner of a game
lose=[]                                                           # final score of loser of a game

def play():
    j=0
    while j<20:                                                    # iterate for 20 games
        cards=[j for i in range(1,5) for j in range(1,14)]         # deck of cards using list comprehension
        np.random.shuffle(cards)                                   # shuffle the deck of cards
        player1=[]
        player2=[]
        k=0
        while k <len(cards):        
            player1.append(cards[k])                              # distribute cards to the players alternatively
            if (k+1<len(cards)):
                 player2.append(cards[k+1])  
            k=k+2
        initial_total1.append(sum(player1))
        initial_total2.append(sum(player2))
        discardpile1=[]
        discardpile2=[]
        i=0
        while i<len(player1):                                 # using while loop to iterate through players cards
            sum1=0                                            # initialize sum1 to 0, sum1 is sum of 2 cards played by player1
            sum2=0
            p1=[]                                             # initialize p1 i.e cards played by player1 
            p2=[]
            p1.append(player1[i])
            p1.append(player1[i+1])
            p2.append(player2[i])
            p2.append(player2[i+1])
            sum1=sum(p1)                                        #sum to find which player has greater sum of cards
            sum2=sum(p2)
            if(sum1>sum2):
                discardpile1.extend(p1)
                discardpile1.extend(p2)
                p1=[]
                p2=[]
                i=i+2
            elif(sum2>sum1):
                discardpile2.extend(p1)
                discardpile2.extend(p2)
                p1=[]
                p2=[]
                i=i+2
            elif(sum1==sum2):
                while(sum1==sum2 and i+2<len(player1)):
                    i=i+2
                    p1.append(player1[i])
                    p1.append(player1[i+1])
                    p2.append(player2[i])
                    p2.append(player2[i+1])
                    sum1=sum1+sum(p1)                                   # add sum of new cards to existing sum1
                    sum2=sum2+sum(p2)
                i=i+2
                if(sum1>sum2):                                           # check if player1 has greater final score with respect to player2, player1 won the round    
                    discardpile1.extend(p1)
                    discardpile1.extend(p2)
                    p1=[]
                    p2=[]
                else:
                    discardpile2.extend(p1)
                    discardpile2.extend(p2)
                    p1=[]
                    p2=[]
        score1=sum(discardpile1)
        score2=sum(discardpile2)
                    
        if(score1>score2):
            win.append(score1)
            win_player1[j]=1
            lose.append(score2)
        elif(score2>score1):
            win.append(score2)
            win_player2[j]=1
            lose.append(score1)
          

      
        j=j+1
    print("Winning scores: ",win)
    print("Losing scores: ",lose)
    
play()

min_of_win=min(win)
mean_of_win=statistics.mean(win)
max_of_win=max(win)
min_of_lose=min(lose)
mean_of_lose=statistics.mean(lose)
max_of_lose=max(lose)

print("The min, mean and max of winning totals is",min_of_win,",",mean_of_win,",", max_of_win,"respectively")
print("The min, mean and max of losing totals is",min_of_lose,",",mean_of_lose, ",",max_of_lose,"respectively")


Winning scores:  [219, 193, 203, 207, 183, 214, 189, 291, 203, 239, 203, 188, 188, 215, 267, 232, 284, 252, 255, 209]
Losing scores:  [145, 171, 161, 157, 181, 150, 175, 73, 161, 125, 161, 176, 176, 149, 97, 132, 80, 112, 109, 155]
The min, mean and max of winning totals is 183 , 221.7 , 291 respectively
The min, mean and max of losing totals is 73 , 142.3 , 181 respectively


#### The mean of winning totals is greater than the mean of losing totals. The minimum of the winning totals is greater than the maximum of losing totals. The sum of winning and losing totals for each game remain constant i.e 364

### Exercise 4.2

In [8]:
win_initial_low=[]                                          # initial total of players that win game with low initial score
total_intial_low_win=0                                      # total players that win game with low initial score
total_intial_high_lose=0                                    # total players that lose game with high initial score
lose_initial_high=[]                                        # initial total of players that lose game with high initial score
for index in range(len(initial_total1)):                    # for loop to iterate through the initial totals list
    if(initial_total1[index]<initial_total2[index]):        # check if initial total of player 1 is less than that of player 2
        if(win_player1[index]==1):                          # if player1 won with low initial score increment total_intial_low_win
            total_intial_low_win=total_intial_low_win+1
            win_initial_low.append(initial_total1[index])
        if(win_player2[index]==0):                          # if player2 lose with high initial score increment total_intial_high_lose
            total_intial_high_lose=total_intial_high_lose+1
            lose_initial_high.append(initial_total1[index])
    elif(initial_total2[index]<initial_total1[index]):      # check if initial total of player 2 is less than that of player 1
        if(win_player2[index]==1):                          # if player2 won with low initial score increment total_intial_low_win
            total_intial_low_win=total_intial_low_win+1
            win_initial_low.append(initial_total2[index])
        if(win_player1[index]==0):                          # if player1 lose with high initial score increment total_intial_high_lose
            total_intial_high_lose=total_intial_high_lose+1
            lose_initial_high.append(initial_total1[index])
print("Total wins with initial low scores",total_intial_low_win)  
proportion_initial_low_wins=total_intial_low_win/len(win)            # proportion of wins with low initial scores
print("Proportion of total wins with initial low scores",proportion_initial_low_wins)
print("Total loses with initial high scores",total_intial_high_lose)  
proportion_initial_high_lose=total_intial_high_lose/len(win)          # proportion of lose with high initial scores
print("Proportion of total loses with initial high scores",proportion_initial_high_lose)
print("lowest initial total to win a game",min(win_initial_low))
print("highest initial total to lose a game",max(lose_initial_high))

Total wins with initial low scores 6
Proportion of total wins with initial low scores 0.3
Total loses with initial high scores 6
Proportion of total loses with initial high scores 0.3
lowest initial total to win a game 158
highest initial total to lose a game 206


### Exercise 4.3

In [9]:
import numpy as np
initial_total_of_winners=[]                          # list of initial totals of winners 
initial_total_of_losers=[]                           # list of initial totals of losers
final_total_of_winners=[]                            # list of final totals of winners
final_total_of_losers=[]                             # list of final totals of losers
for i in range(len(initial_total1)):
    if(win_player1[i]==1):                                               # games which player1 wins
        initial_total_of_winners.append(initial_total1[i])               # append initial total of player 1 for this game to list of initial total of winners        
        final_total_of_winners.append(win[i])                            # append final total of player 1 for this game to list of final total of winners
        initial_total_of_losers.append(initial_total2[i])                # append initial total of player 2 for this game to list of initial total of losers
        final_total_of_losers.append(lose[i])                            # append final total of player 2 for this game to list of final total of losers
    elif((win_player2[i]==1)):                                            # games which player2 wins
        initial_total_of_winners.append(initial_total2[i])                # append initial total of player 2 for this game to list of initial total of winners
        final_total_of_winners.append(win[i])                             # append final total of player 2 for this game to list of final total of winners
        initial_total_of_losers.append(initial_total1[i])                 # append initial total of player 1 for this game to list of initial total of losers
        final_total_of_losers.append(lose[i])                             # append final total of player 1 for this game to list of final total of losers
        
#print(initial_total_of_winners)
#print(initial_total_of_losers)
#print(final_total_of_winners)
#print(final_total_of_losers)
np_initial_total_of_winners=np.asarray(initial_total_of_winners)          # convert all 4 list to numpy array
np_initial_total_of_losers=np.asarray(initial_total_of_losers)
np_final_total_of_winners=np.asarray(final_total_of_winners)
np_final_total_of_losers=np.asarray(final_total_of_losers)

correlation1=np.corrcoef(np_initial_total_of_winners,np_final_total_of_winners)  # correlation between initial total of winners and final total of winners
correlation2=np.corrcoef(initial_total_of_losers,final_total_of_losers)          # correlation between initial total of losers and final total of losers
print("Correlation between initial total of winners and final total of winners")
print(correlation1)
print("Correlation between initial total of losers and final total of losers")
print(correlation2)
        

Correlation between initial total of winners and final total of winners
[[1.         0.66374778]
 [0.66374778 1.        ]]
Correlation between initial total of losers and final total of losers
[[1.         0.66374778]
 [0.66374778 1.        ]]


#### The correlation between initial and final totals depicts that initial and final totals are moderately correlated, i.e it can't be concluded that if a player has higher initial total score then he/she will have higher final score with certainity. This is evident in the above exercise as there is 0.2 proportion of total loses with initial high scores. 
#### Both correlations achieve equal value as in any game the amount with which initial score of winner increases is the same amount with which the initial score of loser decreases. Because the sum of  initial and final totals of winner and losers remain constant 